File: deepseek_proxy.py
"""
Tiny proxy that translates OpenAI Responses API to Chat Completions API.
Codex → this proxy (port 15800) → DeepSeek
"""
import json, sys, threading, re
from http.server import HTTPServer, BaseHTTPRequestHandler
from urllib.request import Request, urlopen
from urllib.error import HTTPError, URLError

TARGET = "https://api.deepseek.com/v1/chat/completions"
API_KEY = "sk-707a90a4206b45e9962d606d7a6434f3"
PORT = 15800

class Handler(BaseHTTPRequestHandler):
    def log_message(self, *a): pass  # silent

    def do_POST(self):
        # Read incoming request
        body_len = int(self.headers.get('Content-Length', 0))
        raw = self.rfile.read(body_len)
        try:
            req = json.loads(raw)
        except:
            self._error(400, "Invalid JSON"); return

        # Debug: log first 500 chars of request
        print(f"[PROXY] {self.path} | {json.dumps(req, ensure_ascii=False)[:500]}")

        # Translate Responses API → Chat Completions
        model = req.get("model", "deepseek-v4-pro")
        max_tokens = req.get("max_output_tokens", 4096)

        # Handle system instructions and input
        messages = []
        instructions = req.get("instructions")
        if instructions:
            messages.append({"role": "system", "content": str(instructions)})

        user_input = req.get("input", "")
        # input can be: string, list of messages, or list of content parts
        if isinstance(user_input, str):
            messages.append({"role": "user", "content": user_input})
        elif isinstance(user_input, list):
            for item in user_input:
                # Item can be a message object with role/content
                if isinstance(item, dict):
                    role = item.get("role", "user")
                    content = item.get("content", "")
                    if isinstance(content, list):
                        # Content can be list of parts (e.g. [{"type":"input_text","text":"..."}])
                        text_parts = []
                        for c in content:
                            if isinstance(c, dict):
                                t = c.get("text", "") or c.get("content", "") or str(c)
                                text_parts.append(t)
                            else:
                                text_parts.append(str(c))
                        content = "\n".join(text_parts)
                    elif not isinstance(content, str):
                        content = str(content)
                    messages.append({"role": role, "content": content})
                elif isinstance(item, str):
                    messages.append({"role": "user", "content": item})

        # Handle conversation history if present
        previous = req.get("previous_response_id")
        if previous:
            # We don't have the history, but this is a continuation
            pass

        # Handle tools (Codex may send tool definitions)
        tools = req.get("tools")
        tool_choice = req.get("tool_choice")

        # Build Chat Completions request
        cc_req = {
            "model": model,
            "messages": messages,
            "max_tokens": max_tokens,
            "stream": False,
        }

        if tools:
            # Map Responses API tools to Chat Completions tools
            cc_tools = []
            for t in tools:
                if isinstance(t, dict) and t.get("type") == "function":
                    cc_tools.append({"type": "function", "function": t})
                elif isinstance(t, dict):
                    cc_tools.append(t)
            if cc_tools:
                cc_req["tools"] = cc_tools
        if tool_choice:
            cc_req["tool_choice"] = tool_choice

        # Forward to DeepSeek
        try:
            data = json.dumps(cc_req, ensure_ascii=False).encode('utf-8')
            r = Request(TARGET, data=data, headers={
                "Authorization": f"Bearer {API_KEY}",
                "Content-Type": "application/json",
            })
            resp = urlopen(r, timeout=120)
            cc_resp = json.loads(resp.read())
        except HTTPError as e:
            err_body = e.read().decode('utf-8', errors='replace')
            print(f"[PROXY] Upstream {e.code}: {err_body[:300]}")
            self._error(e.code, f"Upstream error: {e.reason}"); return
        except URLError as e:
            self._error(502, f"Upstream unreachable: {e.reason}"); return

        # Translate Chat Completions response → Responses API format
        choice = cc_resp.get("choices", [{}])[0]
        msg = choice.get("message", {})
        content = msg.get("content", "")
        if not content:
            content = msg.get("reasoning_content", "")

        output_parts = []
        if content:
            output_parts.append({"type": "output_text", "text": content})

        # Handle tool calls in response
        tool_calls = msg.get("tool_calls", [])
        for tc in tool_calls:
            fn = tc.get("function", {})
            output_parts.append({
                "type": "function_call",
                "id": tc.get("id", ""),
                "name": fn.get("name", ""),
                "arguments": fn.get("arguments", ""),
            })

        responses_reply = {
            "id": cc_resp.get("id", "resp_unknown"),
            "object": "response",
            "model": cc_resp.get("model", model),
            "output": output_parts,
            "usage": {
                "input_tokens": cc_resp.get("usage", {}).get("prompt_tokens", 0),
                "output_tokens": cc_resp.get("usage", {}).get("completion_tokens", 0),
                "total_tokens": cc_resp.get("usage", {}).get("total_tokens", 0),
            },
        }

        if choice.get("finish_reason"):
            responses_reply["status"] = "completed"

        self._json(200, responses_reply)

    def do_GET(self):
        if self.path == "/health":
            self._json(200, {"status": "ok"})
        else:
            self._json(200, {"endpoint": "DeepSeek proxy", "target": TARGET})

    def do_OPTIONS(self):
        self.send_response(200)
        self.send_header("Access-Control-Allow-Origin", "*")
        self.send_header("Access-Control-Allow-Methods", "POST,GET,OPTIONS")
        self.send_header("Access-Control-Allow-Headers", "*")
        self.end_headers()

    def _json(self, code, data):
        body = json.dumps(data, ensure_ascii=False).encode('utf-8')
        self.send_response(code)
        self.send_header("Content-Type", "application/json")
        self.send_header("Access-Control-Allow-Origin", "*")
        self.send_header("Content-Length", str(len(body)))
        self.end_headers()
        self.wfile.write(body)

    def _error(self, code, msg):
        self._json(code, {"error": {"message": msg, "type": "proxy_error"}})

if __name__ == "__main__":
    print(f"DeepSeek proxy: http://127.0.0.1:{PORT} → {TARGET}")
    HTTPServer(("127.0.0.1", PORT), Handler).serve_forever()

File: deepseek_proxy.js
// Proxy: Codex (Responses API) → DeepSeek (Chat Completions)
// Codex sends POST /responses expecting SSE stream
// We translate to DeepSeek Chat Completions, then wrap response in SSE stream
const http = require('http');
const https = require('https');

const TARGET = 'https://api.deepseek.com/v1/chat/completions';
const API_KEY = 'sk-707a90a4206b45e9962d606d7a6434f3';
const PORT = 15800;

function translateRequest(rawBody) {
    const req = JSON.parse(rawBody);
    const model = req.model || 'deepseek-v4-pro';
    const maxTokens = req.max_output_tokens || 16000;

    const messages = [];
    const instructions = req.instructions;
    if (instructions) messages.push({ role: 'system', content: String(instructions) });

    const input = req.input;
    if (typeof input === 'string') {
        messages.push({ role: 'user', content: input });
    } else if (Array.isArray(input)) {
        for (const item of input) {
            if (typeof item === 'string') {
                messages.push({ role: 'user', content: item });
            } else if (item && typeof item === 'object') {
                let role = item.role || 'user';
                // Map unsupported roles
                const validRoles = ['system', 'user', 'assistant', 'tool'];
                if (!validRoles.includes(role)) {
                    role = (role === 'developer') ? 'system' : 'user';
                }
                let content = item.content;
                if (Array.isArray(content)) {
                    content = content.map(c => c.text || c.content || '').join('\n');
                }
                if (content) messages.push({ role, content: String(content) });
            }
        }
    }

    return { model, messages, max_tokens: maxTokens, stream: false };
}

function doRequest(ccReq) {
    return new Promise((resolve, reject) => {
        const postData = JSON.stringify(ccReq);
        const url = new URL(TARGET);
        const opts = {
            hostname: url.hostname, port: 443, path: url.pathname, method: 'POST',
            headers: {
                'Authorization': `Bearer ${API_KEY}`,
                'Content-Type': 'application/json',
                'Content-Length': Buffer.byteLength(postData)
            },
            timeout: 180000
        };
        const upstream = https.request(opts, upstreamRes => {
            let data = '';
            upstreamRes.on('data', chunk => data += chunk);
            upstreamRes.on('end', () => {
                try { resolve(JSON.parse(data)); }
                catch (e) { reject(new Error('Parse error: ' + e.message)); }
            });
        });
        upstream.on('error', e => reject(e));
        upstream.on('timeout', () => { upstream.destroy(); reject(new Error('timeout')); });
        upstream.write(postData);
        upstream.end();
    });
}

const server = http.createServer(async (req, res) => {
    res.setHeader('Access-Control-Allow-Origin', '*');
    res.setHeader('Access-Control-Allow-Headers', '*');
    res.setHeader('Access-Control-Allow-Methods', 'POST,GET,OPTIONS');

    if (req.method === 'OPTIONS') { res.writeHead(200); res.end(); return; }
    if (req.method === 'GET') {
        res.writeHead(200, { 'Content-Type': 'application/json' });
        res.end(JSON.stringify({ status: 'ok' }));
        return;
    }
    if (req.method !== 'POST') { res.writeHead(405); res.end(); return; }

    let body = '';
    req.on('data', chunk => body += chunk);
    req.on('end', async () => {
        const preview = body.length > 200 ? body.substring(0, 200) + '...' : body;
        console.log(`[${new Date().toISOString()}] POST ${req.url} | ${preview}`);

        let ccReq;
        try { ccReq = translateRequest(body); }
        catch (e) {
            res.writeHead(400, { 'Content-Type': 'application/json' });
            res.end(JSON.stringify({ error: { message: e.message } }));
            return;
        }

        const respId = 'resp_' + Date.now() + '_' + Math.random().toString(36).slice(2, 8);

        try {
            const ccResp = await doRequest(ccReq);
            console.log(`[${new Date().toISOString()}] Upstream raw: ${JSON.stringify(ccResp).substring(0, 300)}`);
            const choice = (ccResp.choices || [{}])[0];
            const msg = choice.message || {};
            let fullText = msg.content || '';
            if (!fullText) fullText = msg.reasoning_content || '';

            // SSE headers
            res.writeHead(200, {
                'Content-Type': 'text/event-stream',
                'Cache-Control': 'no-cache',
                'Connection': 'keep-alive'
            });

            // event: response.created
            const created = {
                type: 'response.created',
                response: {
                    id: respId, object: 'response', model: ccResp.model || 'deepseek-v4-pro',
                    status: 'in_progress', output: []
                }
            };
            res.write(`event: response.created\ndata: ${JSON.stringify(created)}\n\n`);

            if (fullText) {
                const itemId = 'msg_' + respId;
                const partId = 'part_' + respId;

                // event: response.output_item.added
                const itemAdded = {
                    type: 'response.output_item.added',
                    output_index: 0,
                    item: {
                        id: itemId,
                        type: 'message',
                        role: 'assistant',
                        status: 'in_progress',
                        content: []
                    }
                };
                res.write(`event: response.output_item.added\ndata: ${JSON.stringify(itemAdded)}\n\n`);

                // event: response.content_part.added
                const partAdded = {
                    type: 'response.content_part.added',
                    item_id: itemId,
                    output_index: 0,
                    content_index: 0,
                    part: {
                        type: 'output_text',
                        text: ''
                    }
                };
                res.write(`event: response.content_part.added\ndata: ${JSON.stringify(partAdded)}\n\n`);

                // event: response.output_text.delta
                const delta = {
                    type: 'response.output_text.delta',
                    item_id: itemId,
                    output_index: 0,
                    content_index: 0,
                    delta: fullText
                };
                res.write(`event: response.output_text.delta\ndata: ${JSON.stringify(delta)}\n\n`);

                // event: response.output_text.done
                const textDone = {
                    type: 'response.output_text.done',
                    item_id: itemId,
                    output_index: 0,
                    content_index: 0,
                    text: fullText
                };
                res.write(`event: response.output_text.done\ndata: ${JSON.stringify(textDone)}\n\n`);

                // event: response.content_part.done
                const partDone = {
                    type: 'response.content_part.done',
                    item_id: itemId,
                    output_index: 0,
                    content_index: 0,
                    part: {
                        type: 'output_text',
                        text: fullText
                    }
                };
                res.write(`event: response.content_part.done\ndata: ${JSON.stringify(partDone)}\n\n`);

                // event: response.output_item.done
                const itemDone = {
                    type: 'response.output_item.done',
                    output_index: 0,
                    item: {
                        id: itemId,
                        type: 'message',
                        role: 'assistant',
                        status: 'completed',
                        content: [{ type: 'output_text', text: fullText }]
                    }
                };
                res.write(`event: response.output_item.done\ndata: ${JSON.stringify(itemDone)}\n\n`);
            }

            // event: response.completed
            const done = {
                type: 'response.completed',
                response: {
                    id: respId, object: 'response', model: ccResp.model || 'deepseek-v4-pro',
                    status: 'completed',
                    output: fullText ? [{
                        id: 'msg_' + respId,
                        type: 'message',
                        role: 'assistant',
                        status: 'completed',
                        content: [{ type: 'output_text', text: fullText }]
                    }] : [],
                    usage: {
                        input_tokens: (ccResp.usage || {}).prompt_tokens || 0,
                        output_tokens: (ccResp.usage || {}).completion_tokens || 0,
                        total_tokens: (ccResp.usage || {}).total_tokens || 0
                    }
                }
            };
            res.write(`event: response.completed\ndata: ${JSON.stringify(done)}\n\n`);
            res.end();

        } catch (e) {
            console.error(`[PROXY ERROR] ${e.message}`);
            res.writeHead(200, { 'Content-Type': 'text/event-stream' });
            const errEvt = JSON.stringify({
                type: 'error',
                error: { message: 'Upstream error: ' + e.message }
            });
            res.write(`event: error\ndata: ${errEvt}\n\n`);
            res.end();
        }
    });
});

server.listen(PORT, '127.0.0.1', () => {
    console.log(`DeepSeek proxy: http://127.0.0.1:${PORT} → ${TARGET}`);
});
